man/man3/semLib.3
.\" Copyright (c) 1999-2005 CNRS/LAAS
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd January 21, 2005
.Dt semLib 3
.Os pocoLibs
.Sh NAME 
.Nm semLib 
.Nd semaphore library for pocoLibs
.Sh SYNOPSIS
.Fd #include <portLib.h>
.Fd "#include <semLib.h>"
.Ft SEM_ID 
.Fn semBCreate "int options"  "SEM_B_STATE initalState"
.Ft SEM_ID 
.Fn semCCreate "int options" "int value"
.Ft SEM_ID
.Fn semMCreate "int options"
.Ft STATUS 
.Fn semDelete "SEM_ID semaphore"
.Ft STATUS
.Fn semGive "SEM_ID semaphore"
.Ft STATUS
.Fn semTake "SEM_ID semId" "int timeout"
.Ft STATUS
.Fn semFlush "SEM_ID semId"
.Sh DESCRIPTION
.Fn semBCreate 
creates a binary semaphore.
.Fn semCCreate 
creates a counting semaphore.
.Fn smMCreate
creates a mutual exclusion semaphore. In the case of a POSIX-style
operating system,  all these three semaphore types
are local to the Unix process that created them. 
.Pp
.Fa options
can be
.Dv SEM_Q_FIFO
to specify that tasks waiting on the semaphore are dequeued in a FIFO
order, or 
.Dv SEM_Q_PRIORITY
to specify that tasks waiting on the semaphore are dequeued by
decreasing priority.
.Pp
.Fa initialState
indicates the initial state of a binary semaphore, either 
.Dv SEM_EMPTY 
or
.Dv SEM_FULL
.Pp
.Fa value
indicates the initial value of a counting semaphore.
.Pp
.Fn semTake
implements the 
.Em P
operation on the semaphore.
.Fa timeout
specifies a number of tick in which the operation must succeed. If the
semaphore was not taken with this delay, 
.Dv ERROR
is returned and the errno value of the task is set to
.Dv S_objLib_TIMEOUT .
.Pp
.Fn semGive
implements the 
.Em V
operation on the given semaphore. 
.Pp
.Fn semFlush
resets the value of a counting or binary semaphore to 0 or 
.Dv SEM_EMPTY ,
causing the next call to
.Fn semTake
to effectively block the calling task.
.Pp
.Fn semDelete
deletes a semaphore from the process and releases any resources used
by the semaphore. 
.Fa semaphore
is the identifier of the semaphore to delete. 
.Pp
.Sh RETURN VALUE
.Fn semBCreate
and
.Fn semCCreate
return the semaphore identifier for the newly created semaphore or
NULL in case of an error. In that case an error code is left in the
task's errno value. 
.Pp
.Fn semDelete ,
.Fn semGive ,
.Fn semTake
and 
.Fn semFlush
return
.Dv OK
or 
.Dv ERROR 
in case an error occured. In that case an error code is left in the
task's errno value.
.Sh ERRORS
The task's errno value can be set to an error code from the operating
system or to:
.Bl -tag -width Er
.It Bq Er S_objLib_OBJ_TIMEOUT
A timeout occured on a
.Fn semTake
operation.
.Sh NOTE
The behaviour of processes blocked on a semaphore when it is deleted is
not defined.
.Sh SEE ALSO
.Xr sem_init 3 ,
.Xr sem_wait 3 ,
.Xr sem_post 3 ,
.Xr pthread_mutex_create 3 ,
.Xr pthread_mutex_lock 3 ,
.Xr pthread_mutex_unlock 3 .


CHANGELOG
Version 2.8, xx, xxxx, xxxx

104. Nuke RCS ids in source files.
103. Call mlockall() in h2initGlob() when running under Xenomai.
102. Go back to pmap_unset() to unregister pre-existing posterServer service.
101. Fixes for building Xenomai support.
100. Fixes for cross-compilation in OpenEmbedded
 99. Display the ntick field of poster's timestamps in posterShow().
 98. Enforce stack size in taskSpawn() using pthread_attr_setstacksize().

Version 2.7, 16 april 2010

 97. Fix updating 'ntick' field in H2TIME timespamps of posters.
     (fallout from change #94, which broke pom-genom).

Version 2.6, 13 April 2010

 96. On Linux systems, let taskSuspend() throw a backtrace to stderr.
 95. Use a pipe to wait until h2dev lock is ready before launching posterServ.
 94. Implement poster time stamps using H2TIMESPEC instead of H2TIME. 
 93. Don't start posterServ if it cannot find h2 devices.
 92. Add regression tests for h2time conversion functions
 91. Constify posterName arguments in posterLib.
 90. A few more fixes for running modules on Darwin.
 89. Add regression tests for h2 semaphores.
 88. Change h2 semaphores to mutexes instead of counting semaphores.

Version 2.5, 6 April 2009

 87. Update shipped autotools files to the version provided by Fedora 10.
 86. Implement poster access statistics.
 85. posterServ: use svc_unregister() instead of pmap_unset().

Version 2.4.1, 9 September 2008

 84. Fix a stack corruption bug introduced by change 83.

Version 2.4, 11 August 2008

 83. Add a 'taskFromThread(char *name)' function. (Matthieu Gallien).
 83. Assume that target systems have posix timers when cross-compiling
 82. Let autoheader generate pocolibs-config.h.in
 81. Switch repository to git
 80. Fix a race in change #77.
 79. Eliminate more uses of sprintf().
 78. Fix tickLib regression test.

Version 2.3, 31 July 2007

 77. Fix taskDelete hooks handling.
 76. Relax constraint on h2timer periods.
 75. h2 init no longer starts posterServ by default. Use h2 init -p.
 74. Add a flag to h2devInit() to control the launch of posterServ.
     Bump portLib major.
 73. Fix build of posterLib tests programs.
 72. Remove VxWorks/RTAI/LXRT parts.

Version 2.2, 19 December 2006

 71. Let configure find the gnu tar program for make dist.
 70. Remove the xes subsystem, which is no longer used.
 69. Preliminary support for building Xenomai 2.x enabled libs, using the
     POSIX skin.
 68. Remove RTAI/LXRT support in configure scripts
 67. Remove use of X11 events in xes_server.
 66. Remove double clear of h2timer counters (Sylvain Joyeux).
 65. Fix build problems on Darwin.
 64. Provide a correct day of week in H2TIME structure.
 63. Fix h2timeShow() that was broken in multiple ways.

Version 2.1, 11 November 2005

 62. Some basic support for kbuild for RTAI 3.1 kernel modules
     (not working).
 61. Rework error messages handling. No more central h2errorList.
 60. More ANSI C prototypes in header files.
 59. Add a autogen.sh script to generate the configure script and garbage
     collect unneeded files.
 58. Merge missing parts of lxrt support in existing files.
 57. Add missing files for lxrt support.

Version 2.1beta1, 26 August 2005

 56. Use vpath <pattern> instead of VPATH in Makefiles to minimize trouble
     when builddir != srcdir and srcdir has been configured already.
 55. Add a man page for h2time functions.
 54. Add a h2GetTimeSpec() function that returs a timespec-like structure.
     This should be used instead of calling clock_gettime() directly.
 53. Check for clock_gettime in configure scripts
 52. Add regression tests.
 51. Packaging fixes.
 50. Add a pkg-config data file and install it
 49. Write manual pages for portLib
 48. return S_symLib_NOT_IMPLEMENTED for symFindByValue()
 47. sprintf() -> snprintf() conversion
 46. less printf()s in portLib
 45. more comments translation
 44. Makefiles cleanups (Sylvain Joyeux, Matthieu Herrb).
 43. Add a h2semSet() function to force the value of a semaphore (Sylvain
     Joyeux).
 42. Add a posterName() function and a FIO_FRESH ioctl to posterLib (Sylvain
     Joyeux).
 41. Valgrind support (Sylvain Joyeux).
 40. Remove some unused files.
 39. Make headers C++ compatible (Sylvain Joyeux).
 38. Add --{disable,enable}-posix-timer option to configure, to be able
     to force the timer type that is used by tickLib, as the auto-detection
     of useable POSIX timer is not always reliable enough.
 37. Added --with-linux-kernel-includes option to specify the path to
     linux kernel headers when compiling for RTAI.
 36. Revert the change that made the posterId in posterServ an uint64_t.
 35. Autotoolize the build of the poster test programs.
 34. Start over with libtool style shared libraries version numbers.
     Bump all major revisions to make sure nothing bad happens.
 33. Add a configure test for usabilility of POSIX timers for a
     100Hz sysClk.
 32. Add a 'tick' regress test to check the accuracy of sysClk.
     (Linux 2.4 POSIX timers are known to be unable to set a clock
     at 100Hz).
 31. Some cleanups in header files: avoid nested includes.
 30. Fix posterLib LP64 issues.
 29. Implemented a real-time logging library for RTAI. Added logInit and
     logEnd. Bumped portLib minor.
 28. Updated RTAI/user-space to handle multi-threaded linux processes.
     It now uses one real-time task per thread.
 27. Added devfs support for RTAI.
 26. install-sh does not overwrite files if they were not modified.
 25. Preliminary RTAI/user-space support. posterLib requires a more
     complicated implementation: as of now, posters can be shared between
     RTAI-tasks and linux processes but this breaks the real-time
     properties of RTAI-tasks.
 24. Added taskDelete() regression test.
 23. Added --with-debug option to configure script. It allows the
     selection of independant components to debug.
 22. Added h2devClean() in comLib. This was previously defined in h2.c
     but it was problematic for user-space RTAI version. Bumped comLib

src/portLib/os/posix/semLib.c
/*
 * Copyright (c) 1999, 2003-2005 CNRS/LAAS
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */
#include "pocolibs-config.h"

#include <errno.h>
#include <pthread.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "portLib.h"
#include "wdLib.h"
#include "errnoLib.h"
#include "tickLib.h"

#include "semLib.h"
const H2_ERROR semLibH2errMsgs[]   = SEM_LIB_H2_ERR_MSGS;

#include <semaphore.h>

#ifdef USE_SEM_OPEN
#define TEMPLATE "/tmp/semaphoreXXXXXXXXXX"
#endif

/* semaphore structure */
struct SEM_ID {
   enum { SEM_T_BIN, SEM_T_CNT, SEM_T_MTX } type;
   union {
      sem_t *sem;
      pthread_mutex_t mtx;
   } v;
#ifndef USE_SEM_OPEN
   sem_t sem_storage;
#else
   char name[32]; /* /semaphoreXXXXXXXXXX */
#endif
};

/*
 * Record errors messages
 */
int
semRecordH2ErrMsgs()
{
    return h2recordErrMsgs("semRecordH2ErrMsg", "semLib", M_semLib, 
			   sizeof(semLibH2errMsgs)/sizeof(H2_ERROR), 
			   semLibH2errMsgs);
}

#ifdef USE_SEM_OPEN
static int
_sem_init(SEM_ID sem, unsigned int value)
{ 
	int fd;

	strlcpy(sem->name, TEMPLATE, sizeof(sem->name));
	fd = mkstemp(sem->name);
	if (fd == -1)
		return -1;
	
	close(fd);
	sem->v.sem = sem_open(sem->name, O_CREAT|O_EXCL, 0700, value);
	if (sem->v.sem == (sem_t *)SEM_FAILED) {
		unlink(sem->name);
		return -1;
	}
	return 0;
}
#endif

/*
 * Create and initialize a binary semaphore 
 */
SEM_ID
semBCreate(int options, SEM_B_STATE initialState)
{
    SEM_ID sem;
    int status;

    sem = malloc(sizeof(struct SEM_ID));
    if (sem == NULL) {
	return NULL;
    }
    sem->type = SEM_T_BIN;
#ifndef USE_SEM_OPEN
    sem->v.sem = &sem->sem_storage;
    status = sem_init(sem->v.sem, 0, initialState);
#else
    status = _sem_init(sem, initialState);
#endif
    if (status != 0) {
        free(sem);
	errnoSet(errno);
	return NULL;
    }
    return sem;
}

/*
 * Create and initialize a counting semaphore 
 */
SEM_ID
semCCreate(int options, int initialCount)
{
    SEM_ID sem;
    int status;

    sem = malloc(sizeof(struct SEM_ID));
    if (sem == NULL) {
	return NULL;
    }
    sem->type = SEM_T_CNT;
#ifndef USE_SEM_OPEN
    sem->v.sem = &sem->sem_storage;
    status = sem_init(sem->v.sem, 0, initialCount);
#else
    status = _sem_init(sem, initialCount);
#endif
    if (status != 0) {
        free(sem);
	errnoSet(errno);
	return NULL;
    }
    return sem;
}

/*
 * Create and initialize a mutex semaphore 
 */
SEM_ID
semMCreate(int options)
{
    SEM_ID sem;
    int status;

    sem = malloc(sizeof(struct SEM_ID));
    if (sem == NULL) {
       return NULL;
    }
    sem->type = SEM_T_MTX;
    status = pthread_mutex_init(&sem->v.mtx, NULL);
    if (status != 0) {
       free(sem);
       errnoSet(status);
       return NULL;
    }

    return sem;
}

/*
 * Destroy a semaphore
 */
STATUS
semDelete(SEM_ID semId)
{
    int status;

    switch(semId->type) {
       case SEM_T_MTX:
         status = pthread_mutex_destroy(&semId->v.mtx);
         break;

       default:
#ifndef USE_SEM_OPEN
         status = sem_destroy(semId->v.sem);
#else
	 sem_unlink(semId->name);
	 status = sem_close(semId->v.sem);
#endif
         break;
    }

    if (status != 0) {
	errnoSet(errno);
	return(ERROR);
    }
    free(semId);
    return OK;
}

/*
 * Give a semaphore (V operation)
 */
STATUS
semGive(SEM_ID semId)
{
    int status;
    
    switch(semId->type) {
       case SEM_T_MTX:
         status = pthread_mutex_unlock(&semId->v.mtx);
         break;

       default:          
	 status = sem_post(semId->v.sem);
         break;
    }
    if (status != 0) {
	errnoSet(errno);
	return(ERROR);
    }
    return(OK);
}

/*
 * Dummy handler for watchdog
 */
static 
void semHandler(int sig)
{
}

/*
 * Take a semaphore (P operation)
 */
STATUS
semTake(SEM_ID semId, int timeout)
{
    WDOG_ID timer;
    unsigned long ticks;
    int status, e;

    switch (timeout) {
      case WAIT_FOREVER:
	while (1) {
	   switch(semId->type) {
	      case SEM_T_MTX:
		 status = pthread_mutex_lock(&semId->v.mtx);
		 break;

	      default:
		 status = sem_wait(semId->v.sem);
		 break;
	   }
	   if (status != 0) {
		if (errno == EINTR) {
		    continue;
		} else {
		    errnoSet(errno);
		    return(ERROR);
		}
	   } else {
	      break;
	   }
	}/* while */
	break;
	
      case 0:
	 switch(semId->type) {
	    case SEM_T_MTX:
	       e = EBUSY;
	       status = pthread_mutex_trylock(&semId->v.mtx);
	       break;

	    default:
	       e = EAGAIN;
	       status = sem_trywait(semId->v.sem);
	       break;
	 }
	 if (status != 0) {
	    if (errno == e) {
		errnoSet(S_semLib_TIMEOUT);
		return ERROR;
	    } else {
		return ERROR;
	    }
	}
	break;
	
      default:
	timer = wdCreate();
	ticks = tickGet();
	wdStart(timer, timeout, (FUNCPTR)semHandler, 0);

	while (1) {
	   switch(semId->type) {
	      case SEM_T_MTX:
		 status = pthread_mutex_lock(&semId->v.mtx);
		 break;

	      default:
		 status = sem_wait(semId->v.sem);
		 break;
	   }
	   if (status != 0) {
		if (errno == EINTR) {
		    if (tickGet() - ticks > timeout) {
			errnoSet(S_semLib_TIMEOUT);
			wdDelete(timer);
			return ERROR;
		    } else {
			continue;
		    }
		} else {
		    errnoSet(errno);
		    wdDelete(timer);
		    return(ERROR);
		}
	    } else {
		break;
	    }
	} /* while */
	wdDelete(timer);
	break;
    } /* switch */
    return OK;
}

/*
 * unblock every task pended on a semaphore 
 */
STATUS
semFlush(SEM_ID semId)
{
   switch(semId->type) {
      case SEM_T_MTX:
	 return ERROR; /* one should not flush a mutex? */
	 break;

      default:
	 while (sem_trywait(semId->v.sem) < 0 && errno == EAGAIN) {
	    sem_post(semId->v.sem);
	 }
	 if (sem_post(semId->v.sem) != 0) {
	    return ERROR;
	 }
	 return OK;
   }
}
